src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+from qibolab.sweeper import Parameter, Sweeper, SweeperType
+
+
+@dataclass
+class DummyPort(Port):
+    """Placeholder.
+
+    Copied over from dummy, may not be required.
+    """
+
+    name: str
+    offset: float = 0.0
+    lo_frequency: int = 0
+    lo_power: int = 0
+    gain: int = 0
+    attenuation: int = 0
+    power_range: int = 0
+    filters: Optional[dict] = None
+
+
+class PulseSimulator(Controller):
+    """Runs quantum dynamics simulation of model of device.
+
+    Interfaces with Qibolab. Useful for testing code without requiring
+    access to hardware.
+    """
+
+    PortType = DummyPort  # Placeholder. Copied over from dummy, may not be required.
+
+    def __init__(
+        self,
+        simulation_config: dict,
+        model_config: dict,
+        sim_opts: Optional[None] = None,
+    ):
+        """Set emulator configuration.
+
+        Args:
+            simulation_config (dict): Simulation configuration dictionary.
+            model_config (dict): Model configuration dictionary.
+            sim_opts (optional): Simulation backend specific object specifying simulation options.
+        """
+        self.simulation_config = simulation_config
+        self.model_config = model_config
+        self.sim_opts = sim_opts
+
+        self.all_simulation_backends = {"Qutip": QutipSimulator}
+
+        self.available_sweep_parameters = {
+            Parameter.amplitude,
+            Parameter.duration,
+            Parameter.frequency,
+            Parameter.relative_phase,
+            Parameter.start,
+        }
+
+        self.update()
+
+    @property
+    def sampling_rate(self):
+        return self.model_config["sampling_rate"]
+
+    def update(self):
+        """Updates the pulse simulator by loading all parameters from
+        `self.model_config` and `self.simulation_config`."""
+        self.simulation_backend_name = self.simulation_config["simulation_backend_name"]
+        self.device_name = self.model_config["device_name"]
+        self.model_name = self.model_config["model_name"]
+        self.emulator_name = f"{self.device_name} emulator running {self.model_name} on {self.simulation_backend_name} backend"
+        self.simulation_backend = self.all_simulation_backends[
+            self.simulation_backend_name
+        ](self.model_config, self.sim_opts)
+
+        self.platform2simulator_channels = self.model_config[
+            "platform2simulator_channels"
+        ]
+
+        self.sim_sampling_boost = self.simulation_config["sim_sampling_boost"]
+        self.instant_measurement = self.simulation_config["instant_measurement"]
+        self.runcard_duration_in_dt_units = self.model_config[
+            "runcard_duration_in_dt_units"
+        ]
+        # self.readout_error = self.model_config["readout_error"]
+        self.readout_error = {}
+        for k, v in self.model_config["readout_error"].items():
+            self.readout_error.update({int(k): v})
```

Comment:
The most idiomatic way would be arguably:
```suggestion
        self.readout_error = {int(k): v for k, v in self.model_config["readout_error"].items()}
```

Reply:
Fixed in 8f01ca1ac5436ce0b9411212d5dca0633d9de5d0.